File: rcppad_examples/rw.R
set.seed(1);
library(MASS)

simdata <- function(){
  local({
    rho=0.9
    sds=seq(0.5,2,length=stateDim)
    sdObs=rep(1,stateDim);
    corrMat=matrix(0.0,stateDim,stateDim)
    for(i in 1:stateDim){
      for(j in 1:stateDim){
        corrMat[i,j] = rho^abs(i-j)
      }
    }
    Sigma=corrMat*(sds %o% sds)
    d=matrix(NA,timeSteps,stateDim)
    obs=d;
    ##init state
    d[1,] = rnorm(stateDim);
    i=1;
    obs[i,] = d[i,] + rnorm(stateDim,rep(0,stateDim),sdObs)
    for(i in 2:timeSteps){
      d[i,] = d[i-1,] + mvrnorm(1,rep(0,stateDim),Sigma=Sigma)
      obs[i,] = d[i,] + rnorm(stateDim,rep(0,stateDim),sdObs)
    }
    matplot(d,type="l")
    matpoints(obs);
  },.GlobalEnv)
}
stateDim=3
timeSteps=100
simdata()
## make ADMB data file
## cat(timeSteps,"\n",file="mvrw.dat");
## cat(stateDim,"\n",file="mvrw.dat",append=TRUE);
## write.table(obs,file="mvrw.dat",append=TRUE,row.names=FALSE,col.names=FALSE)
library(RcppAD)
compile("rw.cpp")
compile("rw_parallel.cpp")
dyn.load("rw.so")
dyn.load("rw_parallel.so")
data <- list(obs=t(obs))
parameters <- list(
  u=data$obs*0,
  transf_rho=0.1,
  logsds=sds*0,
  logsdObs=sdObs*0
  )
newtonOption(smartsearch=FALSE)
obj <- MakeADFun(data,parameters,random="^u",DLL="rw")
obj$hessian <- TRUE

obj$fn()
obj$gr()
system.time(opt <- do.call("optim",obj))
pl <- obj$env$parList() ## <-- List of predicted random effects
matpoints(t(pl$u),type="l",col="blue",lwd=2)
summary(as.mle(opt))


## Permormance analysis in case 1000 timesteps and 10 states
stateDim=10
timeSteps=1000
simdata()
data <- list(obs=t(obs))
parameters <- list(
  u=data$obs*0,
  transf_rho=0.1,
  logsds=sds*0,
  logsdObs=sdObs*0
  )
newtonOption(smartsearch=FALSE)
obj <- MakeADFun(data,parameters,random="^u",DLL="rw_parallel")
ben <- benchmarkParallel(obj,n=1000,cores=1:8)
ben
plot(ben)

File: rcppad_examples/rw.cpp
// Random walk with multivariate correlated increments and measurement noise.
#include <RcppAD.hpp>

/* Parameter transform */
template <class Type>
Type f(Type x){return Type(2)/(Type(1) + exp(-Type(2) * x)) - Type(1);}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_ARRAY(obs); /* timeSteps x stateDim */
  PARAMETER_ARRAY(u); /* State */
  PARAMETER(transf_rho);
  PARAMETER_VECTOR(logsds);
  PARAMETER_VECTOR(logsdObs);
  int timeSteps=obs.dim[0];
  int stateDim=obs.dim[1];
  Type rho=f(transf_rho);
  vector<Type> sds=exp(logsds);
  vector<Type> sdObs=exp(logsdObs);
  // Setup object for evaluating multivariate normal likelihood
  matrix<Type> cov(stateDim,stateDim);
  for(int i=0;i<stateDim;i++)
    for(int j=0;j<stateDim;j++)
      cov(i,j)=pow(rho,Type(abs(i-j)))*sds[i]*sds[j];
  using namespace density;
  MVNORM_t<Type> neg_log_density(cov);
  /* Define likelihood */
  Type ans=0;
  ans-=dnorm(vector<Type>(u(0)),Type(0),Type(1),1).sum();
  for(int i=1;i<timeSteps;i++)    
    ans+=neg_log_density(u(i)-u(i-1)); // Process likelihood 
  for(int i=1;i<timeSteps;i++)
    ans-=dnorm(vector<Type>(obs(i)),vector<Type>(u(i)),sdObs,1).sum(); // Data likelihood
  return ans;
}
